Milestone 2 - improved block interface

dsorber opened the pull request:
These changes extend the previous ("milestone 1") changes to:

-  improve the block interface 
- increase the flexibility for creating custom buffers
- enable device-to-device (D2D) transfers

Commit:
buffer_single_mapped subclasses intended for use with accelerated devices -- initial checkin, WIP

Commit:
enum class used to describe how buffers are used -- initial checkin, WIP

Commit:
Lookup table which maps buffer_types to factory functions for creating new buffer_single_mapped subclasses (for now) -- initial checkin, WIP

Commit:
Added new files to support new buffer interface updates

Commit:
Removed allocate_custom_buffer, free_custom_buffer

Commit:
Updates to support new buffer_single_mapped subclasses

Commit:
Added new files to support new buffer_single_mapped subclasses

Commit:
Updates to support new buffer interface

Commit:
Updates to support new buffer interface, renamed some jump/goto labels

Commit:
High-level cleanup.

Commit:
Additional minor cleanup.

Commit:
Added argument to make_host_buffer() call signature and corrected function body

Commit:
Adjusted get_buffer_type() return type (added reference)

Commit:
Added documentation for two parameters to make_buffer()

Commit:
Temporarily commenting out some functions

Commit:
Fixed make_*_buffer() call signature, including return type

Commit:
Adjusted data type for src_buf_type, est_buf_type (added reference)

Commit:
Adjusted buffer_type_t typedef

Commit:
Adjusted operator<<() call signature

Commit:
work-in-progress

Commit:
Add prototype implementation for storing buffer creation factory
function pointer in the buffer_type object itself.

Commit:
Updated md5 checksum

Commit:
Preparing to remove bufer_type_lookup_table from the source tree

Commit:
Cleaning up merge conflicts and messy commented-out code

Commit:
Added block_sptr argument to ctor to match other factory functions, added call to MAKE_CUSTOM_BUFFER_TYPE

Commit:
Added call to MAKE_CUSTOM_BUFFER_TYPE

Commit:
Added call to MAKE_CUSTOM_BUFFER_TYPE immediately following class declaration

Commit:
Added block_sptr argument to make_buffer_double_mapped() so it matches the other factory function call signatures

Commit:
Reorganize and remove function definitions from header files so added
MAKE_CUSTOM_BUFFER_TYPE calls will compile.

Commit:
Move input_blkd_cb_ready() and input_blocked_callback() implementations
to the buffer class since the functions mostly manipulate buffer members
anyway. The same functions still exist in the buffer_reader class but
now delegate to the buffer class as needed.

Commit:
Abstract the logic of the input_blocked_callback and
output_blocked_callback functions so they can be used on device buffers.
Also being implementing the host_buffer subclass.

Commit:
Finished implementing host_buffer logic as example case to test new
buffer concepts.

Commit:
Implement the cuda_buffer subclass.

Commit:
Fix formatting.

Commit:
Add CUDA memmove implementation using a temporary buffer and two memcpys
for now. It would be worthwhile to experiment with other approaches and
see if they perform better.

Commit:
Remove BUFFER_DEBUG define that was accidentally left in place.

Commit:
Remove lingering dust of the no-longer-needed buffer_type_lookup_table.

Commit:
Change the post_work() function to return void, there is no need for a
return value.

mormj commented on cmake/Modules/GrCUDA.cmake:
```diff
@@ -0,0 +1,17 @@
+# Copyright 2021 Free Software Foundation, Inc.
```
We can drop the FSF for copyright, and you can put Author/BlackLynx

dsorber replied:
Okay. I think that came from copy-paste.  By the way, we did complete "registration" with the FSF for the purposes of contributing to GR.

mormj commented on cmake/Modules/GrCUDA.cmake:
```diff
@@ -0,0 +1,17 @@
+# Setup CUDA
+########################################################################
+
+find_package(CUDA)
```
For the sake of a merge onto master, I'm hesitant to add any CUDA support directly in tree - it might be good to have the custom buffers live outside in the OOT that they will be used.  If we do put them in-tree, there will need to be a lot of conditionals to only compile stuff if it finds CUDA - but then what version, etc.  
CUDA is so complex, that I struggle to think it will be a good idea to have to support it in-tree

dsorber replied:
I wanted to talk to you more about this (and will). As you pointed out CUDA ends up being pretty gnarly dependency so adding it to GR is a non-trivial change.  The benefit to having this stuff in tree is that CUDA OOTs can just point to and use the predefined CUDA buffer.  That said, everything should still work out of tree but each CUDA OOT would need to include a couple of source files that define the CUDA buffer or define their own from scratch, which I can't imagine many folks will want to do.  Perhaps a future version of gr_modtool could automate the inclusion of those files?

mormj commented on gnuradio-runtime/include/gnuradio/block.h:
```diff
@@ -514,6 +517,38 @@ class GR_RUNTIME_API block : public basic_block
+     */
+    void allocate_detail(int ninputs,
+                         int noutputs,
+                         const std::vector<int>& downstream_max_nitems_vec,
```
size_t?

dsorber replied:
Yes.  There seems to be a mixture of types used in various places but many of them probably should be changed to size_t. I can make those updates.

mormj commented on gnuradio-runtime/include/gnuradio/buffer_double_mapped.h:
```diff
@@ -0,0 +1,132 @@
+
+namespace gr {
+
+class vmcircbuf;
```
Does the buffer_double_mapped class always pertain to vmcircbuf, or could it be derived for some other type of double mapped.  If the former, then should things be named or included with the vmcircbuf classes?

dsorber replied:
The buffer_double_mapped class (now a subclass of the abstract base buffer class) always wraps and manages a vmcircbuf.  This is a similar arrangement to how the other buffer subclasses wrap and manage an underlying buffer (or buffers).  For buffer_double_mapped that underlying buffer is the specialized vmcircbuf.  I'm not sure I understand the second half of the question.

mormj commented on gnuradio-runtime/include/gnuradio/cuda_buffer.h:
```diff
@@ -0,0 +1,121 @@
+     * \param buf_owner if the block that owns the buffer which may or may not
+     *                  be the same as the block that writes to this buffer
+     *
+     * The total size of the buffer will be rounded up to a system
```
Does this hold for cuda buffers?

dsorber replied:
Yes. The "buf_owner" relates to the concept of an accelerated block replacing an upstream buffer so that the upstream block uses a buffer subclass that is compatible with the downstream accelerated block.  Keep in mind that the cuda_buffer class actually wraps two underlying buffers: the host buffer and the CUDA device buffer.  Therefore an upstream block running on the host would write to the underlying host buffer, then the runtime will trigger the transfer from the host buffer to the CUDA device buffer, and finally the accelerated CUDA block would read its input data from the underlying CUDA buffer.

mormj commented on gnuradio-runtime/include/gnuradio/hip_buffer.h:
```diff
@@ -0,0 +1,104 @@
+
+namespace gr {
+
+class GR_RUNTIME_API hip_buffer : public buffer_single_mapped
```
What is a hip buffer - perhaps some descriptive comments about what it is used for?

dsorber replied:
HIP is Heterogeneous-compute Interface for Portability, which is AMD's CUDA equivalent.  It is _very_ similar to CUDA so we figured we'd add support for HIP to GR too, to keep the 5-7 people out there that own AMD GPUs (:wink:) happy. The same comments above regarding CUDA in tree inclusion also apply to HIP.  I will add additional comments.

mormj commented on gnuradio-runtime/lib/block.cc:
```diff
@@ -372,11 +380,153 @@ void block::set_min_output_buffer(int port, long min_output_buffer)
+        }
+        GR_LOG_DEBUG(d_logger, msg.str());
+#endif
+        buf = make_buffer(nitems,
```
what is the difference between the try and catch here

dsorber replied:
This is preserved from stock GR.  I believe the idea is that if there is an error with the first allocation attempt you try again and only error out if the second attempt fails.  I'm not sure if the second attempt is (or every has been) meaningful.

mormj commented on gnuradio-runtime/lib/buffer.cc:
```diff
@@ -263,92 +216,57 @@ void buffer::prune_tags(uint64_t max_time)
+void buffer::on_lock(gr::thread::scoped_lock& lock)
 {
-    s_buffer_reader_count++;
+    // NOTE: the protecting mutex (scoped_lock) is held by the custom_lock object
```
This mechanism could probably use some detailed comments on what the purpose is and how it functions

dsorber replied:
Agreed, I will add more detailed comments.  The whole custom lock thing is unfortunately both complex and special purpose.

mormj commented on gnuradio-runtime/lib/buffer.cc:
```diff
@@ -263,92 +216,57 @@ void buffer::prune_tags(uint64_t max_time)
+    if ((d_context == buffer_context::DEFAULT_INVALID) || (d_context == context)) {
+        // Set the context if the existing value is the default or if it is the
+        // same as what's already been set
+        d_context = context;
```
why would it be set multiple times, and that be ok?

dsorber replied:
Context would be set multiple times if an upstream block was connected to multiple downstream blocks.  For now the logic allows context to be set multiple times as long as the value isn't changed.  For now this prevents the case where an upstream host block is connected to both another host block and an accelerated block.  Using this paradigm it might be possible to handle at least some variations of this scenario however I haven't gotten there yet and want to make sure all the basic cases work smoothly before trying out more advanced cases.

mormj commented on gnuradio-runtime/lib/buffer_context.cc:
```diff
@@ -0,0 +1,32 @@
+std::ostream& operator<<(std::ostream& os, const buffer_context& context)
+{
+    switch (context) {
+    case buffer_context::DEFAULT_INVALID:
```
Is this list of contexts generic enough for all types of buffers?  Is ingress/egress a sufficient set of cases that buffers would need to handle?

dsorber replied:
That's a very good question. I think so, but let's discuss if you have other ideas.

mormj commented on gnuradio-runtime/lib/buffer_single_mapped.cc:
```diff
@@ -0,0 +1,403 @@
+    return false;
+}
+
+bool buffer_single_mapped::output_blocked_callback_logic(int output_multiple,
```
Both input and output blocked callback logic should be documented in code to describe how they work and when they are used

dsorber replied:
Agreed, I will add the missing documentation.

Commit:
Begin adding additional documentation in places where it is needed.

Commit:
Make buffer_reader::items_available() const again like it should be.
Also wrap the logger instances added to
block_executor::min_available_space() in the LOG macro so they are only
present if logging is enabled.

Commit:
Remove single_mapped_buffer subclasses from the runtime.  The plan is
for such subclasses to live in external modules.

Commit:
Modify implementation to specify custom buffers per port rather than for
the entire block. This should allow even greater flexibility. Additional
work may be needed on the Python bindings to better support the
buffer_type.

Commit:
Move post_work() calls directly into produce() and produce_each()
functions to handle the case were general_work() returns
WORK_CALLED_PRODUCE, which was previously not handled correctly.

Commit:
Add and correct documentation.

Commit:
Fix previously unseen corner case in
buffer_reader_sm::items_available() .

Commit:
Re-add the host_buffer class.

Commit:
Adding QA test cases for the buffer_single_mapped interface using the
host_buffer class implemenation.

Commit:
Correct formatting issues in test code and fix a few miscellaneous
things.

Commit:
Move memcpy and memmove function pointer typedefs to buffer.h. Also
suppress unused parameter errors in CI as there are numerous places
where unused parameters exist.
